Update the cmake build configuration for AppleClang compiler

JerryShih opened the pull request:
This pr try to merge the https://github.com/pytorch/pytorch/pull/11563 again and fix the linking error in https://github.com/pytorch/pytorch/pull/14837.

JerryShih committed:
Remove the redundant openmp cmake code.

The removed lines are already shown in the same block.

JerryShih committed:
Update detectron openmp option.

The AppleClang compiler needs to link openmp lib explicitly.

JerryShih committed:
Update caffe2 openmp option.

The AppleClang compiler needs to link openmp lib explicitly.

JerryShih committed:
Remove the unused cmake OpenMP link flags.

There is no OpenMP_EXE_LINKER_FLAGS variable in cmake FindOpenMP module.

JerryShih committed:
Update the pytorch OpenMP option.

Use "OpenMP_libomp_LIBRARY" cmake var instead of the hard-coded flag.

JerryShih committed:
Work around the pthread linking problem for asan build.

The libasan contains some pthread functions. So, CMake
assumes that the libasan is the found pthread library.
Then, we will just link to the libasan only. However,
libasan doesn't have the full pthread implementation.
If we try to use some pthread advanced functions, we
will hit the link error.
So, this patch just links the pthread explicitly if the
build is an asan build.

JerryShih committed:
Move the OpenMP checking from ATEN block to "USE_OPENMP" block(just move without any change).

The ATEN has its own OpenMP checking block, and it will also
setup some cmake OpenMP variables if it finds OpenMP. That will
cause the caffe2 to still use OpenMP when we set USE_OPENMP=off.

This patch just move the whole checking codes from from ATEN
to "USE_OPENMP" block. There is "no" additional change.

JerryShih committed:
Increases the linux test timeout to 90min.

The pytorch_linux_xenial_py3_clang5_asan_test test is usually timeout
with 1h setting. Try to increase to 90m instead.

JerryShih committed:
Install libomp from brew for mac build for CI.

JerryShih committed:
Show compiler id in summary.

JerryShih committed:
Show the openmp status for pytorch build.

ezyang commented:
Hmm... GitHub makes this annoying to do, but I'd kind of like it if we could land these patches individually :>

JerryShih commented:
@ezyang
Sorry, I can't catch you. Do you mean I should squash all patches for reviewing?

ezyang commented:
Vice versa actually. But we'll make do.

JerryShih commented:
https://github.com/pytorch/pytorch/pull/15820/commits

I try to split the patches to make each commits clearly. Hope these patches could be reviewed easily.

ezyang commented:
Hi @JerryShih, unfortunately, when I build your branch on CentOS, OpenMP is still disabled. More details coming soon.

JerryShih commented:
> Hi @JerryShih, unfortunately, when I build your branch on CentOS, OpenMP is still disabled. More details coming soon.

@ezyang 
That's odd. I think I just setup the special option/setting for "AppleClang" compiler only. Could you make sure the master code-base has openmp before on CentOS?

This commit [[1]](https://github.com/pytorch/pytorch/pull/15820/commits/0567db3744cc92fa79825ce1a688ea10cccb4136) show the explicit message when the cmake finds openmp successfully for "pytorch". Maybe we could land this trivial patch first to get the openmp status in ci log.

ezyang commented:
Here's the logs from the build: https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a

I confirmed that OpenMP was "disabled" by running

```
>>> import torch
torch.get_num_threads()                                                             
1
```

I'll try and land your first commit, as you suggest.

JerryShih commented:
> Here's the logs from the build: https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a
> 
> I confirmed that OpenMP was "disabled" by running
> 
> ```
> >>> import torch
> torch.get_num_threads()                                                             
> 1
> ```
> I'll try and land your first commit, as you suggest.

@ezyang 
Hold on. I found the message "pytorch is compiling with OpenMP" in https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a
I should trace the torch.get_num_threads(). I don't know the relationship between get_num_threads() and openmp.

ezyang commented:
Yes, my other hypothesis is we might need to rebase to master.

JerryShih commented:
@ezyang 
Ok, I know why you get 1 from torch.get_num_threads().

This python call will call to [[1]](https://github.com/pytorch/pytorch/blob/3b337e7892a7b54758f273477aa092053f02b410/aten/src/TH/THGeneral.cpp#L281),  but I just turn off the openmp for aten library in my patch set. You could test with "USE_OPENMP=ON" in caffe2 cmake [[2]](https://github.com/pytorch/pytorch/blob/3b337e7892a7b54758f273477aa092053f02b410/CMakeLists.txt#L128). If this build config could resolve the problem, I should update the aten cmake config logic.

ezyang commented:
I tested just your commit df7648513cb947ca61bba2cf7308258abdd4a805 and it reported the correct number of threads. So there is definitely something later that is causing the problem.

ezyang commented:
I'm merging the first commit in your series at #16506

JerryShih committed:
Turn on USE_OPENMP by default.

The ATEN should always use openmp if cmake could find the openmp.

In the original code[1], ATEN will add openmp related compiling flags
when it finds openmp. The ATEN openmp checking block is merged into
USE_OPENMP block now. We should turn on USE_OPENMP to build ATEN with
openmp by default.

[1]
https://github.com/pytorch/pytorch/blob/7c66ad7455ae7acf215e07a6d4dcd70736c77517/cmake/Dependencies.cmake#L1131

JerryShih commented:
@ezyang 
Here is the rebase one.
I also turn on the openmp by default in [[1]](https://github.com/pytorch/pytorch/pull/15820/commits/6cc185644714888ca1a4e92410e349bacd001f26).

Please try to test with:
> import torch
torch.get_num_threads()                                                             
1

I don't have a good solution to test the openmp status in pytorch.
In my mind, I could use "add_test()" to add a python openmp test.
After the pytorch build, call "make test"(or ninja test) to trigger this python script.

SsnL commented on torch/CMakeLists.txt:
```diff
@@ -323,11 +325,18 @@ target_link_libraries(torch caffe2_library)
+      target_link_libraries(torch ${OpenMP_libomp_LIBRARY})
+    endif()
+  else()
+    target_link_libraries(torch -fopenmp)
```
I think in general blindly linking `-fopenmp` is a bad idea. Considering the problem with `OpenMP_libomp_LIBRARY` being not available in 3.5 cmake, I would suggest port newer `FindOpenMP.cmake` over, and link against `OpenMP_<lang>_LIBRARIES`.

JerryShih replied:
Update to use ${OpenMP_CXX_LIBRARIES} now.

SsnL commented on tools/build_pytorch_libs.py:
```diff
@@ -181,7 +181,15 @@ def run_cmake(version,
+        # openmp cmake flag
+        # If we setup MACOSX_DEPLOYMENT_TARGET in mac, we should also set the
+        # libomp library flags explicitly.
+        OpenMP_C_FLAGS=os.getenv('OpenMP_C_FLAGS'),
```
Are these really needed? Can't you instead use general purpose flags like `CFLAGS` and `LDFLAGS`?

JerryShih replied:
The change in build_pytorch_libs.py are removed now.

SsnL commented on tools/build_pytorch_libs.py:
```diff
@@ -181,7 +181,15 @@ def run_cmake(version,
+        OpenMP_CXX_FLAGS=os.getenv('OpenMP_CXX_FLAGS'),
+        OpenMP_C_LIB_NAMES=os.getenv('OpenMP_C_LIB_NAMES'),
+        OpenMP_CXX_LIB_NAMES=os.getenv('OpenMP_CXX_LIB_NAMES'),
+        OpenMP_omp_LIBRARY=os.getenv('OpenMP_omp_LIBRARY'))
```
This is not used anywhere. `OpenMP_libomp_LIBRARY` you used below is actually set by `FindOpenMP.cmake`.

JerryShih replied:
These changes are reverted.

ssnl reviewed (commented):
Thanks for doing this in a generally better strategy than what I did :). I have a couple of comments, with the main suggestion being to eliminate the unnecessary extra env flags introduced (i.e., `OpenMP_xxxx`). No user would be able to find them out unless they want to read our CI scripts. Moreover, I think OpenMP is more likely & easier to be used by user if the mechanism can be replaced by 

1. smarter cmake code that test `AppleClang > 7.0` and then set `-Xpreprocessor -fopenmp`, and
2. general purpose flags like `CFLAGS` and `LDFLAG`.

Alternatively, we can also port the new `FindOpenMP.cmake` over, which supports `AppleClang` and discovers `brew` installed locations as long as `/usr/local/lib` is in the search path. :)

JerryShih commented:
@SsnL 
Thank you, I'm trying to have the custom FindOpenMP module in "pytorch/cmake/Modules".

JerryShih committed:
Add custom FindOpenMP cmake module.

This file comes from cmake v3.13.3.
https://github.com/Kitware/CMake/blob/8887ebc69b86606b388272d5c89ce11753bb4804/Modules/FindOpenMP.cmake

JerryShih committed:
Update the custom FindOpenMP cmake module.

1) update the module including expression
e.g
include(${CMAKE_CURRENT_LIST_DIR}/CMakeParseImplicitLinkInfo.cmake)=>include(CMakeParseImplicitLinkInfo)
2) add "/usr/local/opt/libomp" in searching path
Since we might set MACOSX_DEPLOYMENT_TARGET, cmake will just try to search
libomp in CMAKE_OSX_SYSROOT folder which might not include the actual libomp
path. Thus, we add the "brew libomp" default install path in the candidates.
3) show the log for try_compile failed case.

JerryShih commented on caffe2/CMakeLists.txt:
```diff
@@ -224,6 +224,19 @@ if(NOT BUILD_ATEN_ONLY)
+    #https://github.com/Kitware/CMake/blob/42212f7539040139ecec092547b7d58ef12a4d72/Modules/FindOpenMP.cmake#L252
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "7.0" OR
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.0")
+      target_link_libraries(caffe2 PRIVATE ${OpenMP_libomp_LIBRARY})
```
Update to use ${OpenMP_CXX_LIBRARIES} now.

JerryShih committed:
Update the custom FindOpenMP cmake module for old cmake environment.

The "NATIVE_COMMAND" comes from cmake v3.9.6.

JerryShih commented:
@ezyang @SsnL 

Finally, here is the new one.

What I have done here are:
1. add a custom findopenmp module copied from cmake v3.13.
It will check "brew libomp" default install path if we setup MACOSX_DEPLOYMENT_TARGET flag.

2. **use ${OpenMP_CXX_LIBRARIES} in linking setting.**
**This is the risky one. It should be tested with ppc64le platform. I remember that I broke the ppc64le build in my previous patch(https://github.com/pytorch/pytorch/pull/11563). I don't have that platform locally, so I can't test for that platform.**

3. move the openmp checking code from aten code block to "USE_OPENMP" checking block

4. add the hardcoded linking setting for pthread in asan build

5. install libomp from brew for all macos related test builds.

ssnl committed:
Find OpenMP quietly
FindMKL: test found OMP is GOMP before marking GOMP as found

ssnl committed:
display MKL lib and include when found

SsnL commented:
@JerryShih Thanks. The CI failures are due to the tricky case where MKL links a conda `libiomp5` and `FindOpenMP` links a brew one. I'm trying to fix this, and push to your branch.

ssnl committed:
add readme about the ported FindOpenMP.cmake

ssnl committed:
revert the brew include special case; and make FindOpenMP respect QUIET flag

ssnl committed:
Add MKL_OPENMP_TYPE (GNU or Intel) and MKL_OPENMP_LIBRARY outputs to FindMKL.cmake

ssnl committed:
Use policy CMP0025 otherwise XCode C compiler gets detected as Clang,
not AppleClang, at least on my local laptop

ssnl committed:
Patch FindOpenMP.cmake to be compatible with MKL.

ssnl committed:
Add/Prettify various CMake outputs

ssnl committed:
Add test that macOS CI build has OpenMP enabled

SsnL commented:
Locally I've verified that this allows me to use multithreaded reductions like `sum` on CPU.

ssnl committed:
Try MKL libomp generally, not just with AppleClang

SsnL commented:
Will the bad news is that it fails about every CI build with this sccache error:
```
Feb 02 03:26:16 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccAtd869.s: Assembler messages:\n/tmp/ccAtd869.s:72: Error: operand type mismatch for `vxorps\'\n/tmp/ccAtd869.s:76: Error: operand type mismatch for `vandps\'\n/tmp/ccAtd869.s:110: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:113: Error: no such instruction: `kmovb %eax,%k3\'\n/tmp/ccAtd869.s:115: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:116: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:117: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:118: Error: operand type mismatch for `vpxord\'\n/tmp/ccAtd869.s:120: Error: no such instruction: `kmovb %eax,%k4\'\n/tmp/ccAtd869.s:121: Error: operand type mismatch for `vpabsq\'\n/tmp/ccAtd869.s:122: Error: operand size mismatch for `vmovdqa64\'\n" }
```
Any ideas? @ezyang @yf225

ssnl committed:
Fix cmake 3.5 not supporting VERSION_GREATER_EQUAL.

SsnL commented:
Oh I can actually reproduce this on ubuntu. Not on mac tho.

SsnL commented on cmake/Modules/FindMKL.cmake:
```diff
@@ -116,30 +120,52 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES _name _list _flags)
+        ELSEIF(${_library} MATCHES "gomp")
+          SET(_openmp_type "GNU")
+          # Use FindOpenMP to find gomp
+          FIND_PACKAGE(OpenMP QUIET)
```
oh... findmkl and findomp probably should not recursively call into each other :P

ssnl committed:
avoid inifinite recursive calls

ssnl committed:
update readme

ssnl committed:
update note

ssnl committed:
minor typo

ezyang commented:
Oh! When I cleaned and rebuilt the problem went away lol.

SsnL commented:
Pro tip: use ‘—cmake’ when there are canker changes :D
